pkg/kubectl/rolling_updater.go

```diff
@@ -756,13 +749,53 @@ func updateWithRetries(rcClient client.ReplicationControllerInterface, rc *api.R
 			rc = oldRc
 		}
 		// The Get passed: rc contains the latest controller, expect a poll for the update.
-		return false, nil
+		return false
 	})
 	// If the error is non-nil the returned controller cannot be trusted, if it is nil, the returned
 	// controller contains the applied update.
 	return rc, err
 }
 
+type updateFuncForPod func(pod *api.Pod)
+
+// updateWithRetriesForPod updates applies the given pod as an update.
+func updateWithRetriesForPod(podClient client.PodInterface, pod *api.Pod, applyUpdate updateFuncForPod) (*api.Pod, error) {
+	var err error
+	oldPod := pod
+	err = retry(3*time.Second, 3, func() bool {
+		// Apply the update, then attempt to push it to the apiserver.
+		applyUpdate(pod)
+		if pod, err = podClient.Update(pod); err == nil {
+			// pod contains the latest controller post update
+			return true
+		}
+		// Update the pod with the latest resource version, if the update failed we
+		// can't trust pod so use oldPod.Name.
+		if pod, err = podClient.Get(oldPod.Name); err != nil {
+			// The Get failed: Value in pod cannot be trusted.
+			pod = oldPod
+		}
+		// The Get passed: pod contains the latest controller, expect a poll for the update.
+		return false
+	})
+	// If the error is non-nil the returned pod cannot be trusted, if it is nil, the returned
+	// pod contains the applied update.
+	return pod, err
+}
+
+// retry retries fn using exponential backoff
+func retry(delay time.Duration, maxRetries int, fn func() bool) error {
```

Comment:
Take a look at https://github.com/kubernetes/kubernetes/blob/293b0d0815cf8f54ea1df802546b2f61ad4a2a01/pkg/client/unversioned/conditions.go#L67.  We probably want to copy this or move it into a util pkg.